File: apps/nexora-finance/mobile/lib/api/organizations.ts
import type { ApiEnvelope } from './client';
import type { NexoraApi } from './NexoraApiProvider';

export type Organization = {
  id: string;
  name: string;
  slug: string;
  status: string;
};

export type OrganizationMembership = {
  id: string;
  organization_id: string;
  status: string;
};

export async function getOrganizations(request: NexoraApi): Promise<Organization[]> {
  const response = await request.request<Organization[]>('/api/v1/organizations/');
  return requireData(response, 'organizations');
}

export async function getOrganization(
  request: NexoraApi,
  organizationId: string,
): Promise<Organization> {
  const response = await request.request<Organization>(
    `/api/v1/organizations/${encodeURIComponent(organizationId)}/`,
  );
  return requireData(response, 'organization');
}

export async function getOrganizationMembership(
  request: NexoraApi,
  organizationId: string,
): Promise<OrganizationMembership> {
  const response = await request.request<OrganizationMembership>(
    `/api/v1/organizations/${encodeURIComponent(organizationId)}/membership/`,
  );
  return requireData(response, 'organization membership');
}

function requireData<T>(response: ApiEnvelope<T>, resource: string): T {
  if (!response.success || response.data === null) {
    throw new Error(`Nexora Core returned no ${resource} data.`);
  }
  return response.data;
}

File: services/nexora-api/api/v1/organizations.py
from __future__ import annotations

from rest_framework.authentication import BaseAuthentication
from rest_framework.exceptions import PermissionDenied as DRFPermissionDenied
from rest_framework.permissions import IsAuthenticated
from rest_framework.views import APIView

from apps.access.api.permissions import CanReadOrganization
from apps.core.api.response import error_response, success_response
from apps.core.exceptions import NexoraException, PermissionDeniedException
from apps.identity.authentication import FirebaseIdentityAuthentication
from apps.organizations.selectors import get_active_membership, get_user_organizations


class OrganizationAPIView(APIView):
    authentication_classes: list[type[BaseAuthentication]] = [FirebaseIdentityAuthentication]
    permission_classes = [IsAuthenticated]

    def handle_exception(self, exc):
        if isinstance(exc, DRFPermissionDenied):
            return self._error(self.request, PermissionDeniedException("Organization permission is required."))
        return super().handle_exception(exc)

    @staticmethod
    def _error(request, exception: NexoraException):
        return error_response(
            exception.code,
            exception.message,
            details=exception.details,
            correlation_id=getattr(request, "correlation_id", None),
            status=exception.status_code,
        )


class OrganizationListView(OrganizationAPIView):
    def get(self, request):
        organizations = get_user_organizations(request.user)
        data = [
            {
                "id": str(organization.id),
                "name": organization.name,
                "slug": organization.slug,
                "status": organization.status,
            }
            for organization in organizations
        ]
        return success_response(data)


class OrganizationDetailView(OrganizationAPIView):
    permission_classes = [IsAuthenticated, CanReadOrganization]

    def get(self, request, organization_id):
        try:
            membership = get_active_membership(user=request.user, organization_id=organization_id)
            if membership is None:
                raise PermissionDeniedException("Active organization membership is required.")
            organization = membership.organization
            return success_response(
                {
                    "id": str(organization.id),
                    "name": organization.name,
                    "slug": organization.slug,
                    "status": organization.status,
                }
            )
        except NexoraException as exception:
            return self._error(request, exception)


class OrganizationMembershipView(OrganizationAPIView):
    def get(self, request, organization_id):
        try:
            membership = get_active_membership(user=request.user, organization_id=organization_id)
            if membership is None:
                raise PermissionDeniedException("Active organization membership is required.")
            return success_response(
                {
                    "id": str(membership.id),
                    "organization_id": str(membership.organization_id),
                    "role": membership.role.slug.upper(),
                    "status": membership.status,
                }
            )
        except NexoraException as exception:
            return self._error(request, exception)
